### ISSUE TITLE
lmod browser

### ISSUE THREAD

**Issue by johrstrom** [OSC/ondemand#3402]
Lot's of folks are interested in a page that can browse the modules available on a cluster. 

We have this simple class - https://github.com/OSC/ondemand/blob/master/apps/dashboard/app/models/hpc_module.rb. And though it's not an ActiveModel yet, it could be I suppose.

It's not clear to me what one could do with this information inside of OOD, but it seems like they want
* to search modules
* see information about the module they're looking for

**Comment by Xaraxia** [comment]
I would be keen on such a thing. I'm thinking of creating a Passenger app to do exactly this. It would be amazing to have it part of the dashboard (a popup, perhaps, to allow people to browse and then select to fill in a form, but I'm brainstorming here and there might be problematic implications to doing that), but even as a fully-branded Passenger app it would be pretty fantastic.

The challenge is that different partitions potentially have different modules available. They're all in the one shared filesystem, but we alter the modulepath based on architecture. For example, we have AMD epyc3 and epyc4 and Intel Xeon nodes, and these can include a number of different GPUs with mixed architectures (NVIDIA [Hopper, Ampere, Ada] and AMD MI210s so far) and we compile separately for each one for reasons of performance and featureset. So for such a module to work, it would need to be able to filter by architecture. We also split into "local" (hand-built) and "auto" (EasyBuild, but Spack would apply here too). That's less of a problem, because the module path by architecture is known, but it does mean that you can't simply specify a single parent directory per architecture and expect it to work.

I'm not sure what you had in mind, but at the moment I'm looking to generate a JSON per architecture and allow people to select that and then display the module spider output in some sensible fashion. Potentially as a server-side app to avoid sending the entire spider file down to the end-user unnecessarily.

Anyway, I'm keen to help with this one.

**Comment by Xaraxia** [comment]
Oh, incidentally, use-case : because the architectures are all different, but the login nodes are one particular architecture, it's not _possible_ to see the modules for the CUDA nodes from the login node. They are mostly the same, but not entirely, and this trips up a few of our users.

**Comment by Xaraxia** [comment]
Yeah, agree that I wouldn't present the JSON to the end-user. It's just a nice way of storing it at the backend. Module by module works. I'll add it to my TODO list. I don't get a lot of time for OnDemand development, so might be a while, but if I get a chance I'll have a crack at it.

**Comment by hansen-m** [comment]
I think this is a great idea. Could the app/widget just parse the JSON used by auto_modules? I think this would help users find modules to use with the BYOE Jupyter or R apps, or just generally see what is available. Bonus points if they could easily copy/paste from it.

**Comment by simonLeary42** [comment]
https://github.com/UMass-RC/lmod-hierarchy-explorer

**Comment by hansen-m** [comment]
I can share my implementation of this also: https://github.com/hansen-m/module-spider

It exclusively uses the auto_modules .json file (currently only supports a single file) and allows users to search by module name, with a button to copy the module load command to the clipboard.

**Comment by ndjones** [comment]
For a slightly different take on this from an end user docs perspective, this is how we serve this to our users: 
https://docs.nesi.org.nz/Scientific_Computing/Supported_Applications/

this is implemented within a git workflow as part of a statically generated docs site.
the page template is here: https://github.com/nesi/support-docs/blob/main/overrides/supported_apps.html
the javascript behind it is here: https://github.com/nesi/support-docs/blob/main/docs/assets/javascripts/supportedApplications.js
and it is updated based on changes to the listing of modules maintained in this json structure https://github.com/nesi/support-docs/blob/main/docs/assets/module-list.json

**Comment by johrstrom** [comment]
This is coming in 4.1. 

We do have this question for @Xaraxia though - you said:

>  So for such a module to work, it would need to be able to filter by architecture

Do you know what key we can search/filter off of for `architecture`?  We're thinking about dependency modules and displaying them. For example at OSC the `R/4.4.0` module requires `gcc/12.3.0` through the `parentAA` field (that's a list).

**Comment by simonLeary42** [comment]
at my site we have separate module trees for each architecture, so `ml av` on an ARM node shows only ARM modules. My module explorer app takes in a dictionary `{x86_64: /path/to/cache, aarch64: /path/to/cache, ppc64le: /path/to/cache}`

**Comment by Xaraxia** [comment]
@johrstrom Good news on 4.1, thank you! We have something similar to Simon above.

We have /sw/auto/<os>/<architecture> and /sw/local/<os>/<architecture>

architecture is frequently 'noarch' for packages in /sw/local, but everything in /sw/auto is built using EasyBuild, so we have architectures like epyc3, epyc4, epyc3_mi210, epyc4_l40s, xeonsp4_h100. We use gres to target GPUs, and features if they need to target epyc3/epyc4/xeonsp4 specifically. (Usually they don't care, MPI users tend to request a specific one.)

I could walk the tree to generate JSON in whatever format you require as long as we can make the mapping between gres/features and the software packages clear.

**Comment by Bubballoo3** [comment]
Both sub-issues are closed here, do we still need this epic?

**Comment by johrstrom** [comment]
I haven't checked dev yet, but yea I think this is complete and in nightlies now.
